File: backend/app/schemas/decision.py
from pydantic import BaseModel, Field, field_validator
from typing import List, Optional, Dict, Any, Literal
from datetime import datetime

DecisionType = Literal["Architecture", "Research", "Code", "Product", "Career", "Compliance"]
DecisionStatus = Literal["tentative", "confirmed", "revisited", "deprecated"]
DecisionConfidence = Literal["low", "medium", "high"]
DecisionPrivacy = Literal["private", "team", "workspace", "public"]
EffortLevel = Literal["low", "medium", "high"]
RiskLevel = Literal["low", "medium", "high"]
EvidenceSourceType = Literal["chat", "document", "code_run", "external_url", "research_paper"]
EvidenceCredibility = Literal["primary", "secondary", "anecdotal"]
ConstraintType = Literal["budget", "time", "technical", "policy", "team_capacity"]
FlagSeverity = Literal["info", "warning", "critical"]
FlagType = Literal[
    "missing_option",
    "weak_evidence",
    "bias_detected",
    "contradiction",
    "sunk_cost_fallacy",
    "anchoring_bias",
    "availability_bias",
    "groupthink",
    "optimism_bias",
    "status_quo_bias",
    "recency_bias",
    "bandwagon_effect",
]


class AIFlagSchema(BaseModel):
    id: str = Field(min_length=1, max_length=128)
    flag_type: FlagType
    severity: FlagSeverity
    message: str = Field(min_length=1, max_length=500)
    suggested_action: str = Field(min_length=1, max_length=500)
    dismissed: bool = False

class OptionSchema(BaseModel):
    id: str = Field(min_length=1, max_length=128)
    label: str = Field(min_length=1, max_length=200)
    description: str = Field(min_length=1, max_length=2000)
    pros: List[str] = Field(default_factory=list)
    cons: List[str] = Field(default_factory=list)
    estimated_effort: EffortLevel = "medium"
    risk_level: RiskLevel = "low"
    dismissed_reason: Optional[str] = None

class EvidenceSchema(BaseModel):
    id: str = Field(min_length=1, max_length=128)
    source_type: EvidenceSourceType
    source_id: str = Field(min_length=1, max_length=200)
    excerpt: str = Field(min_length=1, max_length=5000)
    credibility: EvidenceCredibility
    added_at: str  # Changed from datetime to str for JSON serialization
    relevance_score: float = Field(ge=0, le=1)

class ConstraintSchema(BaseModel):
    type: ConstraintType
    description: str = Field(min_length=1, max_length=500)
    hard_limit: bool
    current_status: str = Field(min_length=1, max_length=120)

class DecisionBase(BaseModel):
    title: str = Field(min_length=3, max_length=200)
    type: DecisionType
    status: DecisionStatus = "tentative"
    confidence: DecisionConfidence = "medium"
    problem_statement: str = Field(min_length=5, max_length=4000)
    context: Optional[str] = None
    constraints: List[ConstraintSchema] = Field(default_factory=list)
    options: List[OptionSchema] = Field(default_factory=list)
    evidence: List[EvidenceSchema] = Field(default_factory=list)
    tradeoffs: Dict[str, int] = Field(default_factory=dict)
    revisit_rule: Optional[Dict[str, Any]] = None
    ai_flags: List[AIFlagSchema] = Field(default_factory=list)
    tags: List[str] = Field(default_factory=list)
    final_decision: Optional[str] = None
    rationale: Optional[str] = None
    privacy: DecisionPrivacy = "private"

    @field_validator("tradeoffs")
    @classmethod
    def validate_tradeoffs(cls, value: Dict[str, int]) -> Dict[str, int]:
        allowed_keys = {
            "performance",
            "cost",
            "complexity",
            "risk",
            "scalability",
            "time_to_implement",
        }
        for key, score in value.items():
            if key not in allowed_keys:
                raise ValueError(f"Invalid tradeoff key: {key}")
            if not isinstance(score, int) or score < 1 or score > 5:
                raise ValueError(f"Tradeoff '{key}' must be an integer between 1 and 5")
        return value

    @field_validator("tags")
    @classmethod
    def validate_tags(cls, value: List[str]) -> List[str]:
        sanitized = []
        for tag in value:
            cleaned = tag.strip()
            if cleaned:
                sanitized.append(cleaned[:64])
        return sanitized[:30]

class DecisionCreate(DecisionBase):
    workspace_id: Optional[str] = "default"

class DecisionUpdate(BaseModel):
    title: Optional[str] = Field(default=None, min_length=3, max_length=200)
    type: Optional[DecisionType] = None
    status: Optional[DecisionStatus] = None
    confidence: Optional[DecisionConfidence] = None
    problem_statement: Optional[str] = Field(default=None, min_length=5, max_length=4000)
    context: Optional[str] = None
    constraints: Optional[List[ConstraintSchema]] = None
    options: Optional[List[OptionSchema]] = None
    evidence: Optional[List[EvidenceSchema]] = None
    tradeoffs: Optional[Dict[str, int]] = None
    revisit_rule: Optional[Dict[str, Any]] = None
    ai_flags: Optional[List[AIFlagSchema]] = None
    tags: Optional[List[str]] = None
    final_decision: Optional[str] = None
    rationale: Optional[str] = None
    privacy: Optional[DecisionPrivacy] = None

class Decision(DecisionBase):
    id: str
    workspace_id: str
    user_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

File: frontend/src/services/decision.ts
import { useAuthStore } from '@/stores/authStore';

const API_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000/api/v1';

// Ensure API_URL ends with /api/v1
const getBaseUrl = () => {
  let url = API_URL;
  if (!url.includes('/api/v1')) {
    url = url.endsWith('/') ? `${url}api/v1` : `${url}/api/v1`;
  }
  return url;
};

const FINAL_API_URL = getBaseUrl();

export type DecisionStatus = 'tentative' | 'confirmed' | 'revisited' | 'deprecated';
export type DecisionConfidence = 'low' | 'medium' | 'high';
export type DecisionType = 'Architecture' | 'Research' | 'Code' | 'Product' | 'Career' | 'Compliance';
export type DecisionPrivacy = 'private' | 'workspace' | 'public';

export interface Option {
  id: string;
  label: string;
  description: string;
  pros: string[];
  cons: string[];
  estimated_effort: 'low' | 'medium' | 'high';
  risk_level: 'low' | 'medium' | 'high';
  dismissed_reason?: string | null;
}

export interface Evidence {
  id: string;
  source_type: 'chat' | 'document' | 'code_run' | 'external_url' | 'research_paper';
  source_id: string;
  excerpt: string;
  credibility: 'primary' | 'secondary' | 'anecdotal';
  added_at: string;
  relevance_score: number;
}

export interface Constraint {
  type: 'budget' | 'time' | 'technical' | 'policy' | 'team_capacity';
  description: string;
  hard_limit: boolean;
  current_status: string;
}

export interface AIFlag {
  id: string;
  flag_type: 'missing_option' | 'weak_evidence' | 'bias_detected' | 'contradiction' | 'sunk_cost_fallacy' | 
             'anchoring_bias' | 'availability_bias' | 'groupthink' | 'optimism_bias' | 
             'status_quo_bias' | 'recency_bias' | 'bandwagon_effect';
  severity: 'info' | 'warning' | 'critical';
  message: string;
  suggested_action: string;
  dismissed: boolean;
}

export interface RevisitRule {
  trigger_type: 'time_based' | 'metric_based' | 'event_based';
  trigger_value: string;
  notification_enabled: boolean;
}

export interface Decision {
  id: string;
  title: string;
  type: DecisionType;
  workspace_id: string;
  user_id: number;
  created_by?: string;  // Author email, populated by the API from user.email
  created_at: string;
  updated_at: string;
  status: DecisionStatus;
  confidence: DecisionConfidence;
  problem_statement: string;
  context?: string;
  constraints: Constraint[];
  options: Option[];
  evidence: Evidence[];
  tradeoffs: {
    performance: number;
    cost: number;
    complexity: number;
    risk: number;
    scalability: number;
    time_to_implement: number;
  };
  final_decision?: string;
  rationale?: string;
  ai_flags?: AIFlag[];
  revisit_rule?: RevisitRule;
  tags: string[];
  privacy: DecisionPrivacy;
}

const normalizePrivacy = (privacy: unknown): DecisionPrivacy => {
  if (privacy === 'workspace' || privacy === 'public' || privacy === 'private') return privacy;
  if (privacy === 'team') return 'workspace';
  return 'private';
};

const normalizeDecision = (decision: any): Decision => ({
  ...decision,
  privacy: normalizePrivacy(decision?.privacy),
});

export class DecisionAIError extends Error {
  status?: number;

  constructor(message: string, status?: number) {
    super(message);
    this.name = 'DecisionAIError';
    this.status = status;
  }
}

export const decisionService = {
  async getDecisions() {
    const token = useAuthStore.getState().token;
    const response = await fetch(`${FINAL_API_URL}/decisions/`, {
      method: 'GET',
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    });

    if (response.status === 401) {
      throw new Error('Session expired. Please log in again.');
    }
    if (!response.ok) {
      throw new Error(`Failed to load decisions (HTTP ${response.status}).`);
    }
    const data = await response.json();
    if (!Array.isArray(data)) return [];
    return data.map(normalizeDecision);
  },

  async getDecision(id: string): Promise<Decision | null> {
    const token = useAuthStore.getState().token;
    try {
      const response = await fetch(`${FINAL_API_URL}/decisions/${id}`, {
        method: 'GET',
        headers: {
          'Authorization': `Bearer ${token}`,
        },
      });

      if (response.ok) {
        const data = await response.json();
        return normalizeDecision(data);
      }
      return null;
    } catch (error) {
      console.error('Decision service error:', error);
      return null;
    }
  },

  async createDecision(decision: Partial<Decision>, opts?: { idempotencyKey?: string }) {
    const token = useAuthStore.getState().token;
    try {
      const response = await fetch(`${FINAL_API_URL}/decisions/`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`,
          ...(opts?.idempotencyKey ? { 'Idempotency-Key': opts.idempotencyKey } : {}),
        },
        body: JSON.stringify(decision),
      });

      if (response.ok) {
        const data = await response.json();
        return normalizeDecision(data);
      }
      throw new Error('Failed to create decision');
    } catch (error) {
      console.error('Decision service error:', error);
      throw error;
    }
  },

  async updateDecision(id: string, updates: Partial<Decision>) {
    const token = useAuthStore.getState().token;
    try {
      const response = await fetch(`${FINAL_API_URL}/decisions/${id}`, {
        method: 'PATCH',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`,
        },
        body: JSON.stringify(updates),
      });

      if (response.ok) {
        const data = await response.json();
        return normalizeDecision(data);
      }
      throw new Error('Failed to update decision');
    } catch (error) {
      console.error('Decision service error:', error);
      throw error;
    }
  },

  async analyzeDecision(decision: Partial<Decision>): Promise<AIFlag[]> {
    const token = useAuthStore.getState().token;
    try {
      const response = await fetch(`${FINAL_API_URL}/decisions/analyze`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`,
        },
        body: JSON.stringify(decision),
      });

      if (response.ok) {
        return response.json();
      }

      let message = 'Decision analysis failed';
      try {
        const body = await response.json();
        message = body?.detail?.message || body?.detail || message;
      } catch {
        message = `Decision analysis failed with status ${response.status}`;
      }

      throw new DecisionAIError(message, response.status);
    } catch (error) {
      console.error('Decision analysis error:', error);
      if (error instanceof DecisionAIError) {
        throw error;
      }
      throw new DecisionAIError('Decision analysis unavailable. Please retry.', 0);
    }
  },

  async scanWorkspace(): Promise<{ evidence: Evidence[] }> {
    const token = useAuthStore.getState().token;
    try {
      const response = await fetch(`${FINAL_API_URL}/decisions/scan-workspace`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`,
        },
      });

      if (response.ok) {
        return response.json();
      }
      throw new Error('Failed to run codebase scan');
    } catch (error) {
      console.error('Scan workspace error:', error);
      throw error;
    }
  }
};
